proposals/0000-abstract-q.rst

```diff
@@ -0,0 +1,193 @@
+This means that libraries that implement ``Quasi`` instances would likely not have to make any changes if a new method is added.
+
+``qRecover`` cannot be implemented in terms of ``qRun`` as it includes a mention of the monad in negative position.
+
+The rest of the changes are internal to GHC and ``ghc-internal``.
+
+Effect and Interactions
+-----------------------
+
+* The `Pure Template Haskell proposal <https://github.com/ghc-proposals/ghc-proposals/pull/655>`_ aims to
+  empower users to ban use of ``IO`` in Template Haskell splices.
+  This proposal opens up a lightweight implementation path for something along these lines.
+  One could implement a ``dropIO :: Q a -> Q a`` function that removes the ``runIO`` effect from the ``MetaHandlers`` record,
+  replacing it with an error call. This function could only be implemented by accessing ``ghc-internal``.
+
+
+Costs and Drawbacks
+-------------------
+
+The main cost of this proposal is that it entails a breaking change to the ``template-haskell`` interface.
+The implementation should be relatively simple and if anything it should simplify things as an existential is being replaced with a common-or-garden record.
+
+
+Backward Compatibility
+----------------------
+TODO: impact assessment but it's likely to be minor
+
+
+Implementation Plan
+-------------------
+Teo Camarasu will implement this.
+
+This section is mostly here to clarify the internal changes required to implement the changes to the interface.
+
+We would make the following changes in ``GHC.Internal.TH.Syntax``::
+
+  -- we create a new type
+  data MetaHandlers =
+    MetaHandlers
+    { mReify :: Name -> IO Info
+    , mNewName :: String -> IO Name
+    ... and so on
+    }
+
+  -- we change the definition of Q
+  newtype Q a = Q { unQ :: MetaHandlers -> IO a }
+
+  -- we move Quasi Language.Haskell.TH.Syntax
+  -- so the code is deleted from here
+
+We would make the following changes in ``Language.Haskell.TH.Syntax``::
+
+ class (MonadIO m, MonadFail m) => Quasi m where
+  qRun     :: Q a -> m a -- New method
+  qRecover :: m a -> m a -> m a
+  qNewName :: String -> m Name
+  qNewName nm = qRun $ \handlers -> mNewName nm -- we add default methods
+  qReport  :: Bool -> String -> m ()
+  qReport severity msg = qRun $ \handlers -> mReport severity msg -- we add default methods
+  qReify   :: Name -> m Info
+  qReify nm = qRun $ \handlers -> mReify nm -- we add default methods
+  ... and so on
+  {-# MINIMAL qRun qRecover #-}
+
+  instance Quasi Q where
+    qRun = id
+    qRecover r k = catch k (const $ r)
+    -- all other methods are just the default
+
+  runQ :: Quasi m => Q a -> m a
+  runQ = qRun
+
+We would also alter the code for running splices in ``GHC.Tc.Gen.Splice`` and would construct a value of type ``MetaHandlers``.
+
+An important note is that currently the error throwing and recovery logic (``qReport`` and ``qRecover``) for ``Q`` make use of the fact that
```

Comment:
`qReport` doesn't "throw", it only reports. `fail` throws.

Reply:
I've deleted this